crates/lang/src/codegen/dispatch/execution.rs

```diff
@@ -70,24 +88,192 @@ pub struct ExecuteConstructorConfig {
 /// The closure is supposed to already contain all the arguments that the real
 /// constructor message requires and forwards them.
 #[inline]
-pub fn execute_constructor<S, F>(
+pub fn execute_constructor<Contract, F, R>(
     config: ExecuteConstructorConfig,
     f: F,
 ) -> Result<(), DispatchError>
 where
-    S: ink_storage::traits::SpreadLayout,
-    F: FnOnce() -> S,
+    Contract: SpreadLayout + ContractRootKey,
+    F: FnOnce() -> R,
+    <private::Seal<R> as ConstructorReturnType<Contract>>::ReturnValue: scale::Encode,
+    private::Seal<R>: ConstructorReturnType<Contract>,
 {
     if config.dynamic_storage_alloc {
         alloc::initialize(ContractPhase::Deploy);
     }
-    let storage = ManuallyDrop::new(f());
-    let root_key = Key::from([0x00; 32]);
-    push_spread_root::<S>(&storage, &root_key);
-    if config.dynamic_storage_alloc {
-        alloc::finalize();
+    let result = ManuallyDrop::new(private::Seal(f()));
+    match result.as_result() {
```

Comment:
Does it fix the issue https://github.com/paritytech/ink/issues/985? (if you apply this change for `execute_message`)

Reply:
I think I now know what causes failure of result check. It is explained here:
https://github.com/dtolnay/case-studies/blob/master/autoref-specialization/README.md#limitations

Message return values are strictly more powerful than constructor return values therefore this cannot directly applied to fix the bug with messages. However, I am sure we can still fix the result bug with another technique.